R/translate-sql-helpers.R

```diff
@@ -132,19 +132,62 @@ copy_env <- function(from, to = NULL, parent = parent.env(from)) {
 #' @param pad If `TRUE`, the default, pad the infix operator with spaces.
 #' @export
 sql_infix <- function(f, pad = TRUE) {
+  # (x1 <- expr((2 - 1) * x))
+  # #> (2 - 1) * x
+  #
+  # (x2 <- expr(!!expr(2 - 1) * x))
+  # #> (2 - 1) * x
+  #
+  # `x1` and `x2` look the same when printed but their AST (abstract syntax tree)
+  # is different:
+  # lobstr::ast(expr((2 - 1) * x))
+  # #> █─expr
+  # #> └─█─`*`
+  # #>   ├─█─`(`
+  # #>   │ └─█─`-`
+  # #>   │   ├─2
+  # #>   │   └─1
+  # #>   └─x
+  #
+  # lobstr::ast(expr(!!expr(2 - 1) * x))
+  # #> █─expr
+  # #> └─█─`*`
+  # #>   ├─█─`-`
+  # #>   │ ├─2
+  # #>   │ └─1
+  # #>   └─x
+  #
+  # the missing paranthese "(" is an issue when translating infix expression.
+  # This is fixed with `escape_infix_expr()`
+  # see https://github.com/tidyverse/dbplyr/issues/634
   assert_that(is_string(f))
 
   if (pad) {
     function(x, y) {
+      x <- escape_infix_expr(enquo(x))
+      y <- escape_infix_expr(enquo(y))
+
       build_sql(x, " ", sql(f), " ", y)
     }
   } else {
     function(x, y) {
+      x <- escape_infix_expr(enquo(x))
+      y <- escape_infix_expr(enquo(y))
+
       build_sql(x, sql(f), y)
     }
   }
 }
 
+escape_infix_expr <- function(x) {
+  infix_calls <- c("+", "-", "*", "/", "%%", "^")
+  if (quo_is_call(x, n = 2) && call_name(x) %in% infix_calls) {
```

Comment:
I think `quo_is_call(x, name, n = 2)` will work?

And I'm not sure `n = 2` is correct? Doesn't the same problem arise with unary `+` and `-`?

Reply:
Hmm, when do we need parantheses for unary `+` and `-`? I came up with something like

```r
translate_sql(-!!expr(-x))
#> <SQL> --`x`
```

Do you see other cases?

Thanks to your question I just noticed that unary `+` isn't supported at all currently:

``` r
library(dplyr, warn.conflicts = FALSE)
library(dbplyr, warn.conflicts = FALSE)

# works
translate_sql(-2)
#> <SQL> -2.0
# error
translate_sql(+2)
#> Error in eval_bare(x, .env): argument "y" is missing, with no default
```

<sup>Created on 2021-12-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Not sure how/whether we want to handle `translate_sql(-!!expr(-x))`. First the inner part, `-x`, translated to SQL ``-`x` ``. Then the outer minus is translated. Here one would have to analyse the SQL to decide whether parantheses are needed.
Probably it should be sufficient to check if the SQL starts with "-". What do you think?

Reply:
I was thinking of this case:

``` r
dbplyr::translate_sql(-!!rlang::expr(1+2))
#> <SQL> -1.0 + 2.0
```

<sup>Created on 2021-12-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

OTOH is this very common? Should we just suggest that if you're going to interpolate in expressions involving infix operators, you should wrap the whole thing in parentheses to avoid operator precedence issues?